game/screen_songs.hh

```diff
@@ -9,7 +9,7 @@
 #include "video.hh"
 #include "playlist.hh"
 #include "menu.hh"
-#include <aubio/fvec.h>
+#include "aubio/aubio.h"
```

Comment:
Why change this?

Reply:
does not compile actually mainline relies on ffmpeg.hh to pull it, But my patches isolate ffmpeg.h thus here does not compile. I don't remember the exact error, is that really a problem?

Reply:
(FYI still work in progress expect many changes in near future...)

Reply:
I know it might change and it's not really a problem; I'm just curious as to why you're doing some things. I noticed also you got rid of the circular buffer; won't that translate to using more memory?

Reply:
Actually, there is one ring buffer, but that matches better what I need to do. the boost one doe not allow random access which is what I need it. I saw there is a ringbuffer class already in the code, I need to check if I can use it (for now it is a custom implementation). For the memory, actually, I could really drop down the amount needed (I tried 40k during tests) but I kept the value 4320256 as you seem to have chose it carefully.
The main idea in these bunch of patches is to have the decoder thread separated from ffmpeg so that ffmpeg class is just reading and resampling/scaling on demand. I also split ffmpeg in two child classes (for video and audio) which allows to separate clearly things needed on each sides.

Reply:
I see.
The buffer size is set to fit roughly 30 seconds (and a bit extra) which is essentially  all the samples in the preview on the song selection screen. The idea is that the aubio code runs only once per song.

Reply:
However, according to the boost documentation, circular_buffer **does** allow random access iterators, or are you talking about something else?

Reply:
Actually it allows random access int the current range. But here I use completelly arbitrary addresses: with boost, I needed to "fill" all the beginning of ringbuf when "seeking" to the end, moreover, accessing the internal "indexes" is made difficult by the API. I must say I spent quite a lot of time trying to keep it, but I eventually decided this was really overkill and not really adapted to what I was doing (for example, iterators and so on). I end up with something that is really close to what you do when using alsa api, with the read pointer and the write pointer.

Reply:
I guess we'll have to compare performance and memory usage when you're done. 🤷🏼‍♂️. Re: the buffer size, the idea is aubio code runs on song preview load and only once at the beginning (because by nature the tempo algorithm has a latency of samples it needs to process before producing a useful result and also because IMO it's better for the song to take a couple or even several milliseconds more to load than have it stutter as new tempos are calculated). 

I don't mind you shrinking the size if it still works as expected so do feel free to experiment.

Reply:
Yes, for sure. And if you have some specific benches and/or samples with specific "problems" (like the corrupt mp3 of bug #444) I would be happy to test them.

Reply:
I doubt that there is any noticeable performance difference between `std::deque` or a circular buffer. The only real reason to use circular buffers in Performous is to allow lock-free access from PortAudio callbacks, where you are not allowed to wait for anything (things like allocating memory or acquiring a lock will occasionally cause delays that break audio recording and playback).

Reply:
I think this is now ready for reviews.